sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/SimplifySizeExpressionCheck.kt

```diff
@@ -0,0 +1,234 @@
+    definingSupertypes = setOf(packageKotlinCollection, packageKotlinText),
+    name = "count",
+    arguments = emptyArguments
+)
+
+private val isEmptyMatcher = FunMatcher(
+    definingSupertypes = setOf(interfaceKotlinCollection, interfaceKotlinMap, packageKotlinText),
+    name = "isEmpty",
+    arguments = emptyArguments
+)
+
+private val isNotEmptyMatcher = FunMatcher(
+    definingSupertypes = setOf(packageKotlinCollection, packageKotlinText),
+    name = "isNotEmpty",
+    arguments = emptyArguments
+)
+
+private val sizeFieldMatcher = fieldMatcher {
+    fieldMatcher {
+        names = setOf("size")
+        definingSupertypes = setOf(interfaceKotlinCollection, interfaceKotlinMap)
+    }
+    fieldMatcher {
+        names = setOf("length")
+        definingSupertypes = setOf("kotlin.String")
+    }
+}
+
+@Rule(key = "S6529")
+class SimplifySizeExpressionCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = listOf(
+        countMatcher, isEmptyMatcher, isNotEmptyMatcher
+    )
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: ResolvedCall<*>,
+        matchedFun: FunMatcherImpl,
+        kotlinFileContext: KotlinFileContext,
+    ) {
+        if (matchedFun == countMatcher) {
+            checkSizeTest(callExpression, kotlinFileContext)
+        } else {
+            checkIsEmptyTest(callExpression, matchedFun == isEmptyMatcher, kotlinFileContext)
+        }
+    }
+
+    override fun visitReferenceExpression(
+        expression: KtReferenceExpression,
+        kotlinFileContext: KotlinFileContext,
+    ) {
+        if (expression is KtNameReferenceExpression && sizeFieldMatcher.matches(expression, kotlinFileContext.bindingContext)) {
+            checkSizeTest(expression, kotlinFileContext)
+        }
+    }
+
+    private fun checkSizeTest(
+        expression: KtExpression,
+        kotlinFileContext: KotlinFileContext,
+    ) {
+        val sizeTest = expandParenthesis(expandOptionalQualifier(expression.parent)) as? KtBinaryExpression ?: return
+        val operationTokenId = sizeTest.getOperationTokenId()
+        val isSizeEquals = when (operationTokenId) {
+            EQEQ_Id -> true
+            EXCLEQ_Id, GT_Id, LT_Id -> false
+            else -> return
+        }
+
+        val bindingContext = kotlinFileContext.bindingContext
+        if (((operationTokenId != LT_Id) && isIntZeroLiteral(sizeTest.right, bindingContext)) ||
+            ((operationTokenId != GT_Id) && isIntZeroLiteral(sizeTest.left, bindingContext))
+        ) {
+            checkNullTestOrReportSize(sizeTest, isSizeEquals, true, expression.parent as? KtDotQualifiedExpression, kotlinFileContext)
+        }
+    }
+
+    private fun checkIsEmptyTest(expression: KtCallExpression, isIsEmpty: Boolean, kotlinFileContext: KotlinFileContext) {
+        var isEmptyTest: PsiElement = expandParenthesis(expandOptionalQualifier(expression.parent)) as? KtExpression ?: return
```

Comment:
should the inferred type be KtCallExpression?

Reply:
Fixed. This was a relic from an earlier stage of the code. PsiElement is sufficient.